File: libaries/francodb.py
import socket
import struct
from typing import Optional, Tuple, Any, List, Dict, Union

# ==========================================
# 1. CUSTOM EXCEPTIONS
# ==========================================
class FrancoDBError(Exception): pass
class OperationalError(FrancoDBError): pass
class AuthError(FrancoDBError): pass
class QueryError(FrancoDBError): pass

# ==========================================
# 2. PROTOCOL CONSTANTS
# ==========================================
CMD_TEXT   = b'Q'
CMD_JSON   = b'J'
CMD_BINARY = b'B'
HEADER_STRUCT = struct.Struct('!cI')

# ==========================================
# 3. CURSOR CLASS
# ==========================================
class Cursor:
    def __init__(self, connection):
        self._conn = connection
        self._last_result = None
        self.last_raw_bytes = b'' 

    def _recv_n(self, n):
        """Helper: Strictly reads n bytes."""
        data = b''
        while len(data) < n:
            packet = self._conn.sock.recv(n - len(data))
            if not packet:
                raise OperationalError("Connection closed by server")
            data += packet
        self.last_raw_bytes += data
        return data

    def execute(self, fql: str, mode: str = 'text') -> Union[str, List, Dict]:
        if not self._conn.is_connected():
            raise OperationalError("Database is not connected")

        self.last_raw_bytes = b''

        # 1. Select Protocol
        msg_type = CMD_TEXT
        if mode == 'json': msg_type = CMD_JSON
        elif mode == 'binary': msg_type = CMD_BINARY

        # 2. Pack & Send Request
        payload_bytes = fql.encode('utf-8')
        header = HEADER_STRUCT.pack(msg_type, len(payload_bytes))

        try:
            self._conn.sock.sendall(header + payload_bytes)

            # [FIX] 3. Read Response LENGTH Header (4 Bytes)
            # The server now sends the length first!
            len_bytes = self._recv_n(4)
            resp_len = struct.unpack('!I', len_bytes)[0]

            # 4. Read Response BODY
            # We read exactly 'resp_len' bytes. This is the Payload.
            body_data = self._recv_n(resp_len)

            # 5. Parse Payload
            if mode == 'binary':
                return self._parse_binary_body(body_data)
            else:
                return self._parse_text_body(body_data)

        except socket.error as e:
            raise OperationalError(f"Network error: {e}")

    def _parse_text_body(self, data: bytes) -> str:
        """Standard text/json reader."""
        try:
            response = data.decode('utf-8').strip()
            if "ERROR" in response and not response.startswith("{"):
                raise QueryError(response)
            return response
        except UnicodeDecodeError:
            raise OperationalError("Received non-text data")

    def _parse_binary_body(self, data: bytes):
        """
        Unpacks the Professional Binary Protocol from C++.
        Now works on the pre-fetched 'data' buffer.
        """
        try:
            # We need a stream-like object to read sequentially from 'data'
            # Or just use slicing logic. Let's use slicing for clarity.
            ptr = 0
            
            # 1. Read Response Type (1 Byte)
            resp_type = data[ptr]
            ptr += 1

            # --- CASE: ERROR (0xFF) ---
            if resp_type == 0xFF:
                msg_len = struct.unpack('!I', data[ptr:ptr+4])[0]
                ptr += 4
                error_msg = data[ptr:ptr+msg_len].decode('utf-8')
                raise QueryError(f"Server Error: {error_msg}")

            # --- CASE: SIMPLE MESSAGE (0x01) ---
            elif resp_type == 0x01:
                msg_len = struct.unpack('!I', data[ptr:ptr+4])[0]
                ptr += 4
                return data[ptr:ptr+msg_len].decode('utf-8')

            # --- CASE: TABLE DATA (0x02) ---
            elif resp_type == 0x02:
                # A. Read Metadata
                num_cols = struct.unpack('!I', data[ptr:ptr+4])[0]
                ptr += 4
                num_rows = struct.unpack('!I', data[ptr:ptr+4])[0]
                ptr += 4

                # B. Read Columns
                columns = []
                for _ in range(num_cols):
                    # col_type = data[ptr] # Unused in python currently
                    ptr += 1 
                    name_len = struct.unpack('!I', data[ptr:ptr+4])[0]
                    ptr += 4
                    col_name = data[ptr:ptr+name_len].decode('utf-8')
                    ptr += name_len
                    columns.append(col_name)

                # C. Read Rows
                result_rows = []
                for _ in range(num_rows):
                    row_data = []
                    for _ in range(num_cols):
                        val_len = struct.unpack('!I', data[ptr:ptr+4])[0]
                        ptr += 4
                        val_str = data[ptr:ptr+val_len].decode('utf-8')
                        ptr += val_len

                        # Auto-convert numbers
                        if val_str.isdigit() or (val_str.startswith("-") and val_str[1:].isdigit()):
                            row_data.append(int(val_str))
                        else:
                            # Try float
                            try:
                                row_data.append(float(val_str))
                            except ValueError:
                                row_data.append(val_str)

                    result_rows.append(row_data)

                return result_rows

        except Exception as e:
            if isinstance(e, QueryError): raise e
            raise OperationalError(f"Binary parse failed: {e}")

    def close(self): pass
    def __enter__(self): return self
    def __exit__(self, exc_type, exc_val, exc_tb): self.close()

# ==========================================
# 4. CONNECTION CLASS
# ==========================================
class FrancoDB:
    def __init__(self, host='localhost', port=2501, timeout=10):
        self.host = host
        self.port = port
        self.timeout = timeout
        self.sock = None
        self._connected = False

    def connect(self, username='', password='', database=''):
        try:
            self.sock = socket.create_connection((self.host, self.port), timeout=self.timeout)
            self._connected = True
            with self.cursor() as cur:
                if username and password:
                    res = cur.execute(f"LOGIN {username} {password};")
                    if "OK" not in str(res) and "SUCCESS" not in str(res):
                        raise AuthError(f"Login failed: {res}")
                if database:
                    cur.execute(f"2ESTA5DEM {database};")
        except socket.error as e:
            self._connected = False
            raise OperationalError(f"Connection error: {e}")

    def cursor(self): return Cursor(self)
    def is_connected(self): return self._connected and self.sock is not None
    def close(self):
        if self.sock:
            try: self.sock.close()
            except: pass
        self.sock = None
        self._connected = False
    def __enter__(self): return self
    def __exit__(self, exc_type, exc_val, exc_tb): self.close()

def connect(host='localhost', port=2501, user='', password='', database=''):
    conn = FrancoDB(host, port)
    conn.connect(user, password, database)
    return conn

File: libaries/francodb.js
// ==========================================
// 3. CURSOR CLASS (FIXED)
// ==========================================
class Cursor {
    constructor(connection) {
        this._conn = connection;
        this._lastResult = null;
        this.lastRawBytes = Buffer.alloc(0);
    }

    async execute(fql, mode = 'text') {
        if (!this._conn.isConnected()) {
            throw new OperationalError("Database is not connected");
        }

        // 1. Select Protocol
        let msgType = CMD_TEXT;
        if (mode === 'json') msgType = CMD_JSON;
        else if (mode === 'binary') msgType = CMD_BINARY;

        // 2. Pack & Send Request
        const payloadBytes = Buffer.from(fql, 'utf-8');
        const header = Buffer.alloc(5);
        msgType.copy(header, 0); // Byte 0: Type
        header.writeUInt32BE(payloadBytes.length, 1); // Bytes 1-4: Length

        try {
            this._conn.socket.write(Buffer.concat([header, payloadBytes]));

            // 3. Read Response Packet (Length + Body)
            // We use a helper to handle the Length-Prefix framing
            const packet = await this._readPacket();

            // Store for debugging
            this.lastRawBytes = packet;

            // 4. Parse based on Mode
            if (mode === 'binary') {
                return this._parseBinary(packet);
            } else {
                return this._parseText(packet);
            }
        } catch (e) {
            if (e instanceof FrancoDBError) throw e;
            throw new OperationalError(`Network error: ${e.message}`);
        }
    }

    // [FIX] Generic Packet Reader: Reads 4 bytes Length, then N bytes Body
    async _readPacket() {
        return new Promise((resolve, reject) => {
            let buffer = Buffer.alloc(0);
            let needed = 4; // Start by needing 4 bytes for Length Header
            let readingBody = false;

            const timeout = setTimeout(() => {
                cleanup();
                reject(new OperationalError("Query timed out"));
            }, this._conn.timeout);

            const onData = (chunk) => {
                buffer = Buffer.concat([buffer, chunk]);

                // Phase 1: Read Header
                if (!readingBody) {
                    if (buffer.length >= 4) {
                        // We have the length header!
                        needed = buffer.readUInt32BE(0);

                        // Consume header, keep remainder
                        buffer = buffer.subarray(4);
                        readingBody = true;
                    }
                }

                // Phase 2: Read Body (Check repeatedly in case chunk was large)
                if (readingBody) {
                    if (buffer.length >= needed) {
                        // We have the full body!
                        const payload = buffer.subarray(0, needed);
                        cleanup();
                        resolve(payload);
                    }
                }
            };

            const onError = (err) => {
                cleanup();
                reject(err);
            };

            const cleanup = () => {
                clearTimeout(timeout);
                this._conn.socket.removeListener('data', onData);
                this._conn.socket.removeListener('error', onError);
            };

            this._conn.socket.on('data', onData);
            this._conn.socket.on('error', onError);
        });
    }

    _parseText(buffer) {
        const response = buffer.toString('utf-8').trim();
        if (response.includes("ERROR") && !response.startsWith("{")) {
            throw new QueryError(response);
        }
        return response;
    }

    _parseBinary(buffer) {
        let offset = 0;

        // 1. Read Response Type (1 Byte)
        if (buffer.length < 1) throw new OperationalError("Empty binary response");
        const respType = buffer.readUInt8(offset);
        offset += 1;

        // --- CASE: ERROR (0xFF) ---
        if (respType === 0xFF) {
            const msgLen = buffer.readUInt32BE(offset);
            offset += 4;
            const errorMsg = buffer.toString('utf-8', offset, offset + msgLen);
            throw new QueryError(`Server Error: ${errorMsg}`);
        }

        // --- CASE: SIMPLE MESSAGE (0x01) ---
        if (respType === 0x01) {
            const msgLen = buffer.readUInt32BE(offset);
            offset += 4;
            return buffer.toString('utf-8', offset, offset + msgLen);
        }

        // --- CASE: TABLE DATA (0x02) ---
        if (respType === 0x02) {
            const numCols = buffer.readUInt32BE(offset); offset += 4;
            const numRows = buffer.readUInt32BE(offset); offset += 4;

            // B. Read Columns
            const columns = [];
            for (let i = 0; i < numCols; i++) {
                // Skip Type (1 byte)
                offset += 1;
                const nameLen = buffer.readUInt32BE(offset); offset += 4;
                const colName = buffer.toString('utf-8', offset, offset + nameLen);
                offset += nameLen;
                columns.push(colName);
            }

            // C. Read Rows
            const resultRows = [];
            for (let i = 0; i < numRows; i++) {
                const rowData = [];
                for (let j = 0; j < numCols; j++) {
                    const valLen = buffer.readUInt32BE(offset); offset += 4;
                    const valStr = buffer.toString('utf-8', offset, offset + valLen);
                    offset += valLen;

                    // Auto-convert numbers
                    if (/^-?\d+$/.test(valStr)) {
                        rowData.push(parseInt(valStr, 10));
                    } else if (/^-?\d*\.\d+$/.test(valStr)) {
                        rowData.push(parseFloat(valStr));
                    } else {
                        rowData.push(valStr);
                    }
                }
                resultRows.push(rowData);
            }
            return resultRows;
        }

        throw new OperationalError(`Unknown response type: 0x${respType.toString(16)}`);
    }

    close() {
        // No-op
    }
}